pennylane/beta/qnode.py

```diff
@@ -0,0 +1,540 @@
+            the maximum number of derivatives to support. Increasing this value allows
+            for higher order derivatives to be extracted, at the cost of additional
+            (classical) computational overhead during the backwards pass.
+
+    Keyword Args:
+        **kwargs: Any additional keyword arguments provided are passed to the differentiation
+            method. Please refer to the :mod:`qml.gradients <.gradients>` module for details
+            on supported options for your chosen gradient transform.
+
+    **Example**
+
+    QNodes can be created by decorating a quantum function:
+
+    >>> dev = qml.device("default.qubit", wires=1)
+    ... @qml.beta.qnode(dev)
+    ... def circuit(x):
+    ...     qml.RX(x, wires=0)
+    ...     return expval(qml.PauliZ(0))
+
+    or by instantiating the class directly:
+
+    >>> def circuit(x):
+    ...     qml.RX(x, wires=0)
+    ...     return expval(qml.PauliZ(0))
+    >>> dev = qml.device("default.qubit", wires=1)
+    >>> qnode = qml.beta.QNode(circuit, dev)
+    """
+
+    def __init__(
+        self,
+        func,
+        device,
+        interface="autograd",
+        diff_method="best",
+        max_expansion=10,
+        mode="best",
+        cache=True,
+        cachesize=10000,
+        max_diff=1,
+        **gradient_kwargs,
+    ):
+        if interface not in SUPPORTED_INTERFACES:
+            raise qml.QuantumFunctionError(
+                f"Unknown interface {interface}. Interface must be "
+                f"one of {SUPPORTED_INTERFACES}."
+            )
+
+        if not isinstance(device, Device):
+            raise qml.QuantumFunctionError(
+                "Invalid device. Device must be a valid PennyLane device."
+            )
+
+        if "shots" in inspect.signature(func).parameters:
+            warnings.warn(
+                "Detected 'shots' as an argument to the given quantum function. "
+                "The 'shots' argument name is reserved for overriding the number of shots "
+                "taken by the device. Its use outside of this context should be avoided.",
+                UserWarning,
+            )
+            self._qfunc_uses_shots_arg = True
+        else:
+            self._qfunc_uses_shots_arg = False
+
+        # input arguments
+        self.func = func
+        self.device = device
+        self._interface = interface
+        self.diff_method = diff_method
+        self.max_expansion = max_expansion
+
+        # execution keyword arguments
+        self.execute_kwargs = {
+            "mode": mode,
+            "cache": cache,
+            "cachesize": cachesize,
+            "max_diff": max_diff,
+        }
+
+        # internal data attributes
+        self._tape = None
+        self._qfunc_output = None
+        self._gradient_kwargs = gradient_kwargs
+        self._original_device = device
+        self.gradient_fn = None
+        self.gradient_kwargs = None
+

```

Comment:
Oh, good point, yes. 

Early on when I was drafting this new QNode class, I removed the `mutable=...` keyword argument because I realized it is very broken.

In particular, `mutable=False` works a little... too well. It not only freezes the circuit, but also freezes gate arguments 😢 

![image](https://user-images.githubusercontent.com/2959003/133731252-a58f19e1-963b-486b-9db7-fff65f99d0e9.png)

So I couldn't justify porting over an existing feature knowing it is broken...

Reply:
What are your thoughts? Port it over as-is, and prioritize parity of API? Or leave it off, knowing it doesn't work?

Reply:
Oh wow, yeah `mutable` does look a bit too powerful. I like @glassnotes suggestions